arabara/30 second walk and balance

rabaraaaron opened the pull request:
https://sagebionetworks.jira.com/browse/DHP-216
https://sagebionetworks.jira.com/browse/DHP-217

Support for the 30 Second Walk and Walk and Balance assessments.

Commit:
30 second walk assessment. Walk and balance assessment.

Commit:
Using .shared TTS Synthesizer from MobilePassiveData to fix zombie issue. Removed alert from an interrupted assessment due to auto navigation on waiting for timer to run out.

Commit:
iosViewBuilder files

Commit:
Revert "Using .shared TTS Synthesizer from MobilePassiveData to fix zombie issue. Removed alert from an interrupted assessment due to auto navigation on waiting for timer to run out."

This reverts commit e40ed843966b9fc8bd14aa9b7a2da9a54dc8edc5.

Commit:
Fix build issue

Commit:
Changed device for github workflow

rabaraaaron committed:
Update swift.yml

Commit:
Changed minimum target to 15.2 for workflow build

Commit:
Merge branch 'arabara/30-second-walk-and-balance' of https://github.com/Sage-Bionetworks/MotorControl-iOS into arabara/30-second-walk-and-balance

Commit:
Remove alert from an interrupted assessment

Commit:
Fix phone_in_pocket image sizes to match the rest of the Motor Control assessment images

syoung-smallwisdom commented on Sources/MotorControl/Model/WalkingAssessmentObject.swift:
```diff
@@ -0,0 +1,24 @@
+        self.init(identifier: "example", children: [])
+    }
+    
+    /// Maybe need to change nodenavigator in assessment model to handle full instructions
```
?

Might want a todo date/sign?

syoung-smallwisdom commented on Sources/MotorControl/ViewModel/MotionSensorViewModels.swift:
```diff
@@ -188,6 +184,18 @@ final class TappingStepViewModel : MotionSensorStepViewModel {
+}
+
+/// View model for a balance step
+final class BalanceStepViewModel: MotionSensorStepViewModel {
```
What's the difference between these view models that you need separate classes for them?

rabaraaaron replied:
At least one separate class is needed for Walk or Balance steps because Tremor and Kinetic Tremor use the AppBackgroundListener to quit the measure on interruptions, while Walk or Balance steps do not.

syoung-smallwisdom replied:
Makes sense. Or you could make that config?

syoung-smallwisdom commented on Sources/MotorControl/Views/MotorControlAssessmentView.swift:
```diff
@@ -75,6 +75,12 @@ public struct MotorControlAssessmentView : View {
+            }
+            else if let nodeState = state as? BalanceStepViewModel {
+                MotionSensorStepView(state: nodeState)
+            }
```
Again, kinda looks like it's doing the same thing, yeah?

syoung-smallwisdom commented on shared_resources/json/en.lproj/balance_section_step.json:
```diff
@@ -0,0 +1,103 @@
+        "identifier": "motion",
+        "type": "motion",
+        "requiresBackgroundAudio": true,
+        "recorderTypes": ["accelerometer", "gyro", "gravity", "userAcceleration", "attitude", "rotationRate"],
```
If this is ignored, then it should be removed from the JSON.

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+        "recorderTypes": ["accelerometer", "gyro", "gravity", "userAcceleration", "attitude", "rotationRate"],
+        "startStepIdentifier": "motion",
+        "stopStepIdentifier": "motion"
+    }],
```
Again, remove from JSON if not used.

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+        "actions": {
+            "goForward": {
+                "type": "default",
+                "buttonTitle": "Got it"
```
I thought we'd moved to not using custom titles for the buttons and just using "Next"?

rabaraaaron replied:
The other Motor Control json files for Tremor, Kinetic Tremor, and Tapping all had custom forward navigation button titles, but I will delete the 'buttonTitle' field from the json files and use the default 'Next' from AssessmentModel for all Motor Control measures so we can have uniformity.

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+        "actions": {
+            "goForward": {
+                "type": "default",
+                "buttonTitle": "The phone’s volume is up"
```
I thought we'd moved to not using custom titles for the buttons and just using "Next"?

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+        "actions": {
+            "goForward": {
+                "type": "default",
+                "buttonTitle": "Got front pockets"
```
I thought we'd moved to not using custom titles for the buttons and just using "Next"?

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+        "actions": {
+            "goForward": {
+                "type": "default",
+                "buttonTitle": "Putting it in my pocket"
```
I thought we'd moved to not using custom titles for the buttons and just using "Next"?

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+            "type": "singleColor",
+            "colorStyle": "white"
+        },
+        "subtitle": "This test measures your leg movement and you will need to put your phone in your front pocket. If you do not have pockets you can place the phone in the waistband of your pants.",
```
Should this be "detail"? Seems more like a detail than a subtitle.

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+            "placementType": "fullsizeBackground"
+        },
+        "duration": 5,
+        "commands": ["playSoundOnStart", "transitionAutomatically", "speakWarningOnPause"],
```
Again, are these commands used or ignored? If ignored, should remove from JSON.

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+            "skip": {
+                "type": "navigation",
+                "skipToIdentifier": "countdown",
+                "buttonTitle": "Restart test"
```
We definitely don't support this custom button title. Remove it.

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_section_step.json:
```diff
@@ -0,0 +1,135 @@
+            "reviewInstructions": {
+                "type": "navigation",
+                "skipToIdentifier": "walkInstructions",
+                "buttonTitle": "Review instructions"
```
Same with this one. Not supported. Remove from the JSON.

syoung-smallwisdom reviewed (changes requested):
Suggest reviewing the JSON for stuff that's in there that is not supported or used as configurable code.

Commit:
Addressing PR comments regarding json and redundant code for Walk and Balance and 30 Second Walk steps. Formatting adjustments to Walk and Balance and 30 Second Walk json files.

syoung-smallwisdom commented on Sources/MotorControl/ViewModel/MotorControlAssessmentStates.swift:
```diff
@@ -16,26 +16,16 @@ final class MotorControlAssessmentViewModel : AssessmentViewModel {
-        else if let step = node as? WalkNodeObject {
-            return WalkStepViewModel(step,
+        else if let step = node as? MotionSensorNodeObject {
+            return TremorStepViewModel(step,
```
? If this view model is used for all motion sensor nodes, maybe a rename?

syoung-smallwisdom commented on Sources/MotorControl/Views/MotorControlAssessmentView.swift:
```diff
@@ -67,19 +67,13 @@ public struct MotorControlAssessmentView : View {
-            else if let nodeState = state as? BalanceStepViewModel {
+            else if let nodeState = state as? MotionSensorStepViewModel {
                 MotionSensorStepView(state: nodeState)
+                    .modifier(AppBackgroundListener())
```
Is the app background listener suppose to be there for all? or just if *not* putting the phone in your pocket?

syoung-smallwisdom commented on shared_resources/json/en.lproj/balance_section_step.json:
```diff
@@ -1,103 +1,76 @@
+                    "actions": {
+                        "goForward": {
+                            "type": "default"
+                        }
```
Just remove the button. A default button should always have either a title or icon. Same goes for other cases where you are just deserializing an empty button.

syoung-smallwisdom commented on shared_resources/json/en.lproj/kinetic_tremor.json:
```diff
@@ -27,7 +27,7 @@
                     "goForward":{
                         "type":"default",
-                        "buttonTitle":"Get Started"
+                        "buttonTitle":"Get started"
```
It's a title so grammatically, it should be "Get Started".

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_and_balance.json:
```diff
@@ -1,71 +1,71 @@
+                    "actions":{
+                        "goForward":{
+                            "type":"default",
+                            "buttonTitle":"Get started"
```
```suggestion
                            "buttonTitle":"Get Started"
```

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_thirty_second.json:
```diff
@@ -1,61 +1,60 @@
 {
-    "identifier":"Walk30Seconds",
+    "identifier": "Walk30Seconds",
```
these identifiers should match the assessment identifier.

syoung-smallwisdom commented on shared_resources/json/en.lproj/walk_thirty_second.json:
```diff
@@ -1,61 +1,60 @@
+                    "actions":{
+                        "goForward":{
+                            "type":"default",
+                            "buttonTitle":"Get started"
```
```suggestion
                            "buttonTitle":"Get Started"
```

syoung-smallwisdom reviewed (changes requested):
Bunch more nits. Looks close. :)

Commit:
Combined Walk step and Balance step view model logic to differentiate from Tremor and Kinetic Tremor for the AppBackgroundListener

syoung-smallwisdom commented on Sources/MotorControl/Model/MotionSensorNodeObject.swift:
```diff
@@ -32,18 +32,21 @@ class TremorNodeObject : MotionSensorNodeObject {
 }
 
+class WalkOrBalanceNodeObject : MotionSensorNodeObject {
+}
```
Feels a bit more like where you might use a flag (or protocol or generic) that than checking cast and having these be different types?

Commit:
PR changes. Fixes to all step json files. Removed TremorStepViewModel to just use MotionSensorStepViewModel instead.

syoung-smallwisdom committed:
Do not update view state on a background thread

The backgrounding notification status change was being updated on a
background thread. Need to wrap the handler.

syoung-smallwisdom committed:
Remove json elements that are not implemented and simplify decoding

syoung-smallwisdom committed:
Fix text to speech and phone screen lock handling

syoung-smallwisdom committed:
Update package to fix iOS 16 crashes

syoung-smallwisdom committed:
Change from requiresBackgroundAudio to allowScreenLock

The tremor test should cancel if the app is backgrounded (phone locked) but *not* if the phone goes to sleep due to inactivity.

syoung-smallwisdom committed:
Bump AssessmentModel to decode spoken instruction on completion step

syoung-smallwisdom committed:
Merge pull request #58 from Sage-Bionetworks/syoung/30-second-walk-and-balance

Recommended changes

syoung-smallwisdom committed:
Fix empty buttons

syoung-smallwisdom reviewed (approved):
Looks good
